### ISSUE TITLE
UTF-8 NFD file name on SMB storage cannot be accessed

### ISSUE THREAD

**Issue by PVince81** [owncloud/core#21365]
### Steps

1. Setup a SMB server on Linux
1. Configure a share "shared"
1. Directly inside the share's folder on disk (not going through the network), create a file "good-ümlaut.txt"
1. Create another file "bad-ümlaut.txt"
1. Convert the file name encoding to NFD: `convmv -r -f utf8 -t utf8 --nfd --notest bad-ümlaut.txt` (ref: http://laufer.tumblr.com/post/1066670961/solved-weird-problem-with-special-characters-in).
1. Setup an external storage in OC and mount that share as "/smb"
1. Navigate into the share in the web UI

### Expected result
File appears in list

### Actual result
File not visible

```
{"reqId":"KAz39vteBSxGCMxr1TEG","remoteAddr":"127.0.0.1","app":"OC\\Files\\Cache\\Scanner","message":"!!! Path 'encoding\/bad-u\u0308mlaut.txt' is not accessible or present !!!","level":0,"time":"2015-12-24T09:58:29+00:00","method":"GET","url":"\/owncloud\/index.php\/apps\/files\/ajax\/list.php?dir=%2Fsmb%2Fencoding&sort=name&sortdirection=asc"}
```

### Versions
Observed in OC 8.2.2

**Comment by PVince81** [comment]
Had a chat with @icewind1991 and he told me that the file names are normalized by OC.
So if the remote SMB server contains a file with a file with a "mac" umlaut, it will normalize the name and store the normalized name in the file cache. Then later when it wants to download the file, it would use the normalized name. So obviously that name doesn't exist on the SMB storage => not found.

We cannot disable normalization of file names because it will break syncing on Mac OS then.

Ideal would be if the SMB server itself would have a setting that tells it to normalize UTF-8 names...

So far the only known workaround is to use `convmv` recursively to fix the file names on the remote share.

@icewind1991 @DeepDiver1975 @butonic won't fix?
CC @cmonteroluque

**Comment by PVince81** [comment]
Another idea would be to have ownCloud detect such files and automatically normalize the file names on the remote storage, but that would be a bit "intrusive" and it doesn't feel like it's OC's job to do that.

**Comment by PVince81** [comment]
More info about NFD vs NFC:
- https://en.wikipedia.org/wiki/Filename#Unicode_interoperability

**Comment by PVince81** [comment]
Another possibly, very ugly and with possible performance impact: whenever we have a UTF-8 file name and connect to the remote SMB, try with both the NFC and NFD file name encoding. If one matches, then assume it's the same file.
But then what happens if the SMB storage has two files, one with each encoding ? :boom:

**Comment by PVince81** [comment]
I just tried locally and it is **indeed** possible to have both files with the same name and different encoding:
```
-rw-r--r-- 1 vincent nobody 18 Dec 24 10:51 bad-ümlaut.txt
-rw-r--r-- 1 vincent nobody 18 Dec 24 11:24 bad-ümlaut.txt
-rw-r--r-- 1 vincent nobody 32 Dec 24 10:51 good-ümlaut.txt
```

My font represent them slightly differently:
![umlauts](https://cloud.githubusercontent.com/assets/277525/11993428/288624f6-aa31-11e5-8d3c-f837d4d7fe23.png)

**Comment by PVince81** [comment]
We could also reopen the discussion about adding a new column to oc_filecache to store the real file name (maybe call it `storage_name`).
Let me find the old ticket about it.

**Comment by PVince81** [comment]
also CC @nickvergessen who worked on this idea back then

**Comment by PVince81** [comment]
Somehow can't find an exact ticket.
Found some bits here:
- Discussion about case sensitive remote FS: https://github.com/owncloud/core/issues/17161#issuecomment-115197250 referring to https://github.com/owncloud/core/wiki/Cross-Platform-File-Handling#platform-name
- similar issue but involving the client: https://github.com/owncloud/core/issues/12044

**Comment by PVince81** [comment]
- the PR that removed the file_mapper table with bits of the extra column topic: https://github.com/owncloud/core/pull/17379#issuecomment-118378230

Setting to 9.0 for now to keep in view.

**Comment by karlitschek** [comment]
@DeepDiver1975 FYI For the WND discussions

**Comment by PVince81** [comment]
Can we add a new column in oc_filecache to store the original non-normalized file name ?
@DeepDiver1975 @karlitschek @nickvergessen

**Comment by PVince81** [comment]
and @icewind1991

**Comment by nickvergessen** [comment]
> Can we add a new column in oc_filecache

technically or in terms of "should we".
With a millions of entries the migration step is going to take hours, so I'd rather use another way when possible.

**Comment by PVince81** [comment]
Both.

I don't see another solution for this, except maybe having this value in a separate table, mapped by file id. If we don't store the non-normalized name there is no way to properly match it on the remote storage.

**Comment by PVince81** [comment]
@icewind1991 can you post information about NFC/NFD collisions, in what direction they can happen and how often ?

**Comment by PVince81** [comment]
Discussed so far: if we'd add an additional column in the cache, we'd need to query that column in the Storage layer which would cause additional SQL queries and impair performance.
So basically if ownCloud calls `$storage->file_exists($someFile)` with $someFile being the normalized name (NFC), the storage would have to ask the cache again to find out what the original NFD name was. This would cause additional queries.

**Comment by PVince81** [comment]
@icewind1991 said it might be possible to provide an mount options to tell the storage to **always** manually convert all file names to NFD when accessing the storage. But this assumed that **ALL** files on the remote storage are using the NFD encoding, and none is using NFC.

**Comment by PVince81** [comment]
I was wondering why we are normalizing the file names in the first place, the answer is that it is needed due to sync clients and also due to the fact that URLs must be encoded in NFC format. Not doing so would then break the web UI and Webdav (and sync clients since they use the Webdav URLs)

**Comment by PVince81** [comment]
Regarding the proposal from https://github.com/owncloud/core/issues/21365#issuecomment-173637039 with the extra column: since the extra DB query to retrieve the non-normalized name can penalize the performance, one idea is to make it optional, per storage.
So whenever someone sets up an external storage, it's disabled by default.
If NFD files are to be used, the admin can choose to enable the option in the mount options dropdown.
This would impact performance, but only in such specific cases.

It could be implemented as a storage wrapper that only gets inserted when the option is set.
The storage wrapper would retrieve the non-normalized name from the database.

**Comment by PVince81** [comment]
@karlitschek @icewind1991 @DeepDiver1975 let me know what you think about https://github.com/owncloud/core/issues/21365#issuecomment-173641544
So far I think it's the best solution.

**Comment by LukasReschke** [comment]
@icewind1991 Might https://github.com/owncloud/core/pull/19825 help here a bit?

**Comment by karlitschek** [comment]
Hmmm. Sounds interesting.  But would needs testing.
@DeepDiver1975 What do you think?

**Comment by DeepDiver1975** [comment]
Sounds reasonable - and this also adds testing effort - (not only but as well) on ci.

We basically need to run our smb tests twice - so 4 in total:
- windows and samba
- with NFD and without NFD support

**Comment by PVince81** [comment]
I don't think https://github.com/owncloud/core/pull/19825 will work.
Because when we call "file_exists($nfcName)", how can we find the nfd name to compare against ?
We'd have to do a `ls` on the remote SMB first to get ALL file names and then try and find the matching name in the list. => even worse performance

**Comment by PVince81** [comment]
Yeah, looks like this PR is already doing this with `readdir`: https://github.com/owncloud/core/pull/19825/files#diff-c167e2f5bf992ec3582fe3c11d63ffa0R72
Might work on a local storage but could be slow on remote storages.

**Comment by CaptionF** [comment]
@PVince81 @cmonteroluque @karlitschek @DeepDiver1975, @bboule   What is the decision ? Prospect is going to make a decision tomorrow.  Will we get this ready in the next maintenance 8.2.3?  Or 9.0 ?  The sooner the better!  what is the release date of 8.2.3 ?

**Comment by LukasReschke** [comment]
> what is the release date of 8.2.3 ?

https://github.com/owncloud/core/wiki/Maintenance-and-Release-Schedule

**Comment by PVince81** [comment]
Whichever the decision, the change will be too big to make it into a maintenance release.
So it is most likely that whichever solution is implemented makes it either into 9.0 or 9.1.

@karlitschek @DeepDiver1975 @icewind1991 any comments on the proposed solutions above ? Either use @LukasReschke's very slow readdir solution (https://github.com/owncloud/core/issues/21365#issuecomment-173873526) or the less slow "query the database again for the non-normalized name but only when a ext storage flag is explicitly set" (https://github.com/owncloud/core/issues/21365#issuecomment-173641544)

**Comment by PVince81** [comment]
in the case of @LukasReschke's solution it would be better to also have an optional flag to have this slow process disabled by default

**Comment by PVince81** [comment]
For easier testing, I made this ZIP file: https://s3.owncloud.com/owncloud/index.php/s/c8gosiF8GZZHYMT
Extract it directly on a SMB storage and then mount it with ownCloud.

Regardless of the solution, one of the tricky parts is to not break ownCloud whenever the file name exists with BOTH encodings at the same time in the same folder. Such case is provided in the ZIP file as well.

**Comment by PVince81** [comment]
I had a closer look and tested @LukasReschke's PR and it turns out that all it does is force-normalize NFD file names when accessing the directory. I misunderstood it initially and it actually will not help at all with reverse name lookup / denormalizing.

Let me summarize the possible discussed solutions so far in my next post.

**Comment by PVince81** [comment]
0) Solution 0:
    - Idea: have the users run a script *once* on their external storages that automatically renames the NFD files to NFC.
   - Pro: no code changes needed
   - Con: only suitable for situations where the files were created there by mistake
   - Con: renaming remote files might break other software that access the SMB storage outside of OC (and except the files to not be renamed)

1) Solution 1:
   - Idea: have ownCloud automatically rename NFD files to NFC on the remote storages at scan time
   - Pro: no further changes required nor performance impact during regular access
   - Con: renaming the files remotely could break other software that access the SMB storage outside of OC (like for solution 0)

2) Solution 2:
   - Idea: store the non-normalized name in a new nullable column in the "oc_filecache" table to be able to convert it back
   - Detail: for every remote storage access, a storage wrapper will convert the current file name to the non-normalized name by doing an additional database lookup. This lookup will reduce the performance, so this option is *disabled by default*. The option could be called "increase file name compatibility for this storage (affects performance)"
   - Pro: smaller code change that somewhat fits the architecture
   - Pro: works with situations where two files with the same NFC+NFD name exists on the remote storage (corner case)
   - Con: additional database column
   - Con: performance impact at run time (can be mitigated by making it optional)
   - [x] Con: slower upgrade when adding a column to the database: @karlitschek already confirmed that there is no performance impact (at least with MySql)
   - [ ] Research required: do all database engines support storing NFD-style file names ?
   - [ ] Research required: @icewind1991 to be able to read the non-normalized name from the DB, the storage wrapper needs access to the file id. Is it available ?

3) Solution 3:
   - Idea: for every remote storage access, check for both the NFC and NFD file name
   - Con: performance impact as every file_exists scan has to be done twice to the remote storage
   - Con: doesn't support having both NFC + NFD same file name on remote storage (corner case)
   - Con: limited to NFC<->NFD transformations (in case there are other normalized cases we don't know of yet)
   - [ ] Con: @icewind1991 pointed out that NFC<->NFD mapping is not always one to one, to be confirmed

4) Solution 4:
   - Idea: when calling `file_exists($nfcName)` on the remote storage, grab the whole folder contents, normalize all names and then try to find a matching entry in the list
   - Con: heavy performance impact because we need to fetch the whole directory entry often
   - Con: same issues as with solution 3

Reminder: this problem only exists for situations where the SMB storage is accessed by Mac clients **outside** of ownCloud. For people using SMB exclusively through ownCloud, there are no NFD issues.

Did I miss anything ?

**Comment by karlitschek** [comment]
@PVince81 Great summary as discussed.
@CaptionF Does this help you with the customer conversation.
Option 2,3 and 4 requires significant development time and are most likely not back-portable to 8.2
Option 0 and 1 still requires time and significant testing but could _maybe_ done for 8.2

**Comment by CaptionF** [comment]
yes, shared with prospect. they would also accept a script to rename. 
For other accounts, with environments with plenty of Macs, this may not be feasible. So a general solution would be required.

**Comment by PVince81** [comment]
@CaptionF here is the workaround for solution 0:

1. Mount the SMB share locally on a Linux box to get shell access, for example as "/mnt/smb" (or alternatively get shell access on the target machine directly if it's a Linux box)
1. Install the "convmv" tool if not present (ref: 
1. Do a dry run of `convmv` to find out which files it would repair: `convmv -r -f utf8 -t utf8 --nfc /mnt/smb/`. This will show a list of files with the NFD encoding (the problematic one) that the tool would repair by converting it to NFC encoding (the good one).
1. After checking that list, do the real run by adding `--notest` to the command: `convmv -f utf8 -t utf8 --nfc -r --notest /mnt/smb/`
1. Optional: run `occ files:scan` for that storage to make sure it rescans these files

**Comment by CaptionF** [comment]
@PVince81 great !  Thanks a lot!

**Comment by butonic** [comment]
@PVince81 Looking at the umlaut scenario, what does a mac actually do? rename an NFC file to NFD or save it along the other file? If it actually renames the file I would just try to also find the NFD version if the NFC version could not be found.

From a user perspective the chance of wanting to store two files, one NFD, the other NFC at the same time seems vastly smaller than trying to access a file ... no matter the encoding. The end user couldn't care less.

I wonder how NFD encoded files land on the windows share in the first place. If I create a file on HFS+ it will be NFD. Always. So copying it to a windows share probably stores it there as NFD as well. What if I create a file directly on the windows share. I assume Mac OS will use NFD as well.

We should just try to find NFD encoded files if the NFC encoded path did not work. This will make things a little slower if a file is not found, but will give users the expected behavior in 99.999999999% of all cases. Wanting to store the same file in NFD and NFC is a theoretical exercise IMO. Before checking for a NFD version we can compare the two path versions to see if they actually differ.

That being said converting NFD to NFC is IMHO the same thing to do because only apple uses NFD on their HFS+ filesystem. They can open NFC on a windows share just fine. Windows and Linux users will be pretty happy if we fix all the broken filenames in their network drives for them.

If we don't want to mako the choice we could give it to the administrators: search for NFD if NFC is not found (performance hit) or automatically convert NFD files to NFC.

Until we implement anything we can recommend the workaround in https://github.com/owncloud/core/issues/21365#issuecomment-176688606

**Comment by MorrisJobke** [comment]
> I wonder how NFD encoded files land on the windows share in the first place.

I just checked it: I created a NFD encoded file, checked with the convmv tool, then put it into a ownCloud client synced folder and rechecked on the server. All is fine. On the server NFC is used.

**Comment by butonic** [comment]
summoning @MTRichards, @owncloud/designers, @karlitschek, @DeepDiver1975 for a decision on where to take this.

**Comment by PVince81** [comment]
If we can confirm that NFC<->NFD conversions are always one to one (or at least in most sane use cases), then I'd vote for solution 3: https://github.com/owncloud/core/issues/21365#issuecomment-176638080 implemented as an optional storage wrapper. This means the admin must manually enable an additional mount option for that storage, knowing that this will cause a small performance impact.
The storage wrapper (a bit like a storage plugin) will then make sure to check both NFC + NFD file name for every file operation.

**Comment by MTRichards** [comment]
From my side, I think @butonic said it best: 
>The end user couldn't care less.

It just needs to work. If this is the best approach (find an NFD file if NFC isn't found) that makes it seamless, ok.

Question is, does the admin even know this is an issue - and should it just be on all the time for SMB external storage servers? Always there, but skipped if the NFC file is found...would that have a performance impact?

**Comment by PVince81** [comment]
> Always there, but skipped if the NFC file is found...would that have a performance impact?

Good point. In most cases it should automatically find that the NFC file exists.

The cases where it will be slower is:
1. when the sync client does an "existence check". Before uploading, the Android client usually does a HEAD call on the file name to see whether it exists before uploading. That check would cause 2 calls instead of one to the SMB if the file does not exist (first NFC, then NFD). Not sure how bad that would be... 
2. whenever ownCloud tries to access a file that has been deleted remotely from the SMB, then it would have to do two `file_exists` checks before realizing the file is gone. Might not be too bad.

Maybe it would be fine to have it enabled by default, but still leave it as an option for the admin to disable if they are sure that their storage is only used through OC and/or they are sure they will never have NFD (mac) files there.

@icewind1991 can you comment ?

**Comment by karlitschek** [comment]
I agree with the summary of @PVince81 We should do this with the proposed switch but I'm not sure about the default. My original idea was to have it off by default. But could be discussed.

**Comment by icewind1991** [comment]
I would go with off by default

**Comment by PVince81** [comment]
The milestone is still set to 9.0, does it mean we agree to implement this as a bugfix in the context of 9.0 ? Do we agree on the implementation and should we move forward with the implementation ?
@cmonteroluque @karlitschek @DeepDiver1975 

Technically it would be an additional storage wrapper that doubles every file operation (once for NFC and if fails, again for NFD), the storage wrapper can be disabled as a mount option.

**Comment by karlitschek** [comment]
I think the architecture is clear.
But I don't know the temperature of the issue and if someone has time to do the work

**Comment by cmonteroluque** [comment]
If the customer situation is at hand (with the script) and we can move it forward, I would propose this goes to 9.1

**Comment by CaptionF** [comment]
@cmonteroluque yes, it is: #https://github.com/owncloud/enterprise/issues/971
so yes 9.1 
any instance with macs storing files directly in external files will create such issue!

**Comment by PVince81** [comment]
@MTRichards @cmonteroluque @karlitschek something for the 9.1 list ?

**Comment by PVince81** [comment]
@CaptionF any update about the workaround ? Did it work ?

**Comment by MTRichards** [comment]
Yep. I already linked it over.

**Comment by karlitschek** [comment]
I would only do this if the current hotfix is not enough.

**Comment by MTRichards** [comment]
>I would only do this if the current hotfix is not enough. 

Got it. So we need to know ASAP whether this hotfix does or does not work.

**Comment by MTRichards** [comment]
@CaptionF ?

**Comment by CaptionF** [comment]
@MTRichards That patch is not sufficient. Many employees of propsect work always with MACs, so the script workaround approach would need to run the whole day. Still it would take to long for the files to be available for owncloud.  The prospect not happy and waiting for the solution.

**Comment by MTRichards** [comment]
Let me know which prospect please, that would be helpful (out of band of course).

**Comment by cmonteroluque** [comment]
I think it needs a proper fix, I agree. 9.1 is the right timeline

**Comment by PVince81** [comment]
As discussed, use the storage wrapper solution 3 from https://github.com/owncloud/core/issues/21365#issuecomment-176638080

**Comment by SergioBertolinSG** [comment]
This is happening also in google drive. Maybe all external storages?

**Comment by butonic** [comment]
sure ... whenever something changes the encoding outsido of oc we can run into this problem.

Maybe it would make more sense if OC and the clients ignored the UTF encoding and used it as is ... AFAICT Win/Mac/Linux understand both ... not sure where the translation happens. IIRC the client does some magic to always talk NFC to the server .... @dragotin care to shed some light into the reason?

**Comment by dragotin** [comment]
@butonic no, we can not ignore that problem as it is not true that the OSes understand both. If it were, we would not have this problem.

The client normalizes all filenames when they were read from a native system to NFC unicode. This happens with this function: https://github.com/owncloud/client/blob/master/csync/src/std/c_string.c#L263

This very carefully engineered (don't ask where my hair color comes from ;-) function converts from win32 wide char (a problem that the server does not have as it does not run on windows) and from NFD utf8 on mac (within the macro `WITH_ICONV`). There is also a function that does the other way round, from NFC UTF8 to locale. 

That is the exact same problem solved that you guys experience on the server now, with a little difference: The client *knows* that it is always NFD when he's on Mac. In the server external storage case it can be NFC ("usual" case) or NFD which makes solving it a bit more complicated.

I think solution 3) is fine for the error case that we see now: Whenever the file is accessed on the storage, first the NFC name is checked, and if that fails, NFD is tried. A problem is if the file needs to be written later: The correct file name needs to be known for that. You could, again, first try the NFC, if that fails, try NFD, however, here you would have to know if the file is new or not, new files should probably default to NFC.

I think another thing is very important: As on the client, I would demand that in ownCloud core, every filename is supposed to be in UTF8 NFC. That saves a lot of trouble in there. All file system access happens through the external storage api, and always happen with NFC filenames from the core POV. 

The conversions have to happen in the external storage layer. Whenever a filename is read from the external storage (through `readdir` usually), the name needs to be converted to NFC immediately with a function similar to the client ones. That requires that core does not access the files directly, I hope that is doable.

For file access (stat(), write(), open() and such), core continues to use NFC based names, and the specific external storage implementation needs to know how to convert the file name to access the right file. That can either happen through trail and error (try NFC first, if fail, try NFD), which might be tricky for the cases where the file may simply not exist. The other alternative is that the ext storage module somehow remembers the encoding of the filenames in the DB, which is prolly better.

Bottom line: _Keep core NFC only_. Use the ext storage implementations as converter. That is what @DeepDiver1975 and me discussed ages ago already.

@Danimo fyi, that fun again, did I say right?

**Comment by MTRichards** [comment]
Estimation: 1 week

**Comment by butonic** [comment]
OS X uses HFS+ and by default stores files in that encoding. I wonder what happens when you mount a windows share (real windows, not samba) in OS X and store a file with umlaut on it. Will windows see the correct character? will it be stored as NFC / NFD? Is there some magic conversion implemented on windows or macos? I don't have a mac to test this, but looking at the reports from our customers it seems that the filename is stored as is (NFD) and the windows clients should see broken chars. If that is the case we can at least point in that direction and blame apple...

**Comment by MorrisJobke** [comment]
00004418

**Comment by MorrisJobke** [comment]
00004611

**Comment by dragotin** [comment]
I mounted a share from my windows machine on my Mac and put some *interesting* files on it. 
![bildschirmfoto 2016-04-14 um 11 42 03](https://cloud.githubusercontent.com/assets/1070214/14532144/574a107c-0260-11e6-9948-6432dcb5ec34.png)

**Comment by dragotin** [comment]
And that ends up on windows like this:
![glyphs](https://cloud.githubusercontent.com/assets/1070214/14532338/01ca6a1a-0261-11e6-9ede-009a389f2c26.png)

So there is no blaming. Lets find out how the filenams are encoded now on windows.

**Comment by dragotin** [comment]
All files synced nicely from Windows to ownCloud, with proper names. That makes me confident that the file names on windows are properly encoded.

**Comment by MorrisJobke** [comment]
I had another case. There the files were migrated from an old setup. So it's possible that they were already broken on that old setup, because the storage is only used by ownCloud directly.

**Comment by MorrisJobke** [comment]
> I had another case. There the files were migrated from an old setup. So it's possible that they were already broken on that old setup, because the storage is only used by ownCloud directly.

00005213

**Comment by danimo** [comment]
@MorrisJobke occ scan:files needs to warn about NFD files, at least on local storage.

**Comment by PVince81** [comment]
PR to display a warning when scanning on the CLI: https://github.com/owncloud/core/pull/24341
(preliminary work)

**Comment by PVince81** [comment]
Okay, I did some debugging in regards to normalized file names and found this interesting behavior:
- whenever the scanner finds a NFD file name, it keeps the non-normalized name most of the time and only normalizes at the time it writes it to the database
- the entry is correctly inserted into the database
- however, there is a call to `Watcher::cleanFolders()` which will iterate over every file in the folder to automatically remove all entries where `file_exists` is false. Since `file_exists` for the non-normalized name it will remove the entry.

@icewind1991 now thinking of it, why do we even have `cleanFolders` ? I thought `scanChildren` was already able to do a diff of the directory contents. Maybe something to consider separately for perf improvements.

**Comment by PVince81** [comment]
- extra challenge: moving NFD file to trash and back again, the file name would be normalized on restore. I think that should be fine

**Comment by PVince81** [comment]
WIP PR for the new storage wrapper here: https://github.com/owncloud/core/pull/24349

**Comment by PVince81** [comment]
Raised https://github.com/owncloud/core/issues/24421 to discuss `cleanFolder()`
